### ISSUE TITLE
Abort when training simple model on Vega FE

### ISSUE THREAD

**Issue by MJDSys** [ROCmSoftwarePlatform/tensorflow-upstream#1016]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Centos 7 with latest kernel-ml release, using Ubuntu 18.04 docker image with ROCm 3.5
- TensorFlow installed from (source or binary): pip/official docker
- TensorFlow version (use command below):
- Python version: v2.2.0-30-g34c3143 2.2.0
- ROCm version: 3.5.0
- GPU model and memory: Vega FE 16G

**Describe the current behavior**
When running the script below, python will also abort when running the fit function.  Occasionally a memory fault is reported as well.  Changing the b/s variables to be of smaller sizes will allow the script to proceed.  This failure happens everytime on my gpu, but runs fine using upstream tensorflow on a cpu.  This happens before the first epoch completes.

**Describe the expected behavior**
The script should run to completion.

**Standalone code to reproduce the issue**
```python3
#!/usr/bin/python3

import tensorflow as tf
from tensorflow.keras import layers, Model, optimizers, regularizers
import numpy as np

a = 1
b = 17
s = 16


i = tf.keras.Input(shape=(a,))

layer = layers.Dense(b)
normalization = layers.BatchNormalization()

out = normalization(layer(i))

x = np.random.rand(s,a)
y = np.random.rand(s, out.shape.as_list()[1])

m = tf.keras.Model(inputs=i, outputs=out)

m.compile('nadam', 'mse')
m.fit(x, y, epochs=100)
```
[link](https://gitlab.mjdsystems.ca/snippets/1)

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
[Log with LOG_LEVEL=3 + serialization](https://gitlab.mjdsystems.ca/snippets/2)

**Comment by jerryyin** [comment]
Think this one will need your help/context again @deven-amd . I see `normalization_test` get disabled in [this PR](https://github.com/tensorflow/tensorflow/pull/38289/files#diff-3eee62da1db136b74c9730124ff9bbc9R629). Do you remember the context of it/ why it fails? We can suggest the user to fall-back to an older docker image if this is ROCm3.5 related.

**Comment by deven-amd** [comment]
possibly ROCm3.5 related.

The testcase runs fine for me with TF built from source (on `develop-upstream` branch)

**Comment by jerryyin** [comment]
I can confirm that something is going wrong. In my case the behavior is different. I used `rocm/tensorflow-autobuilds:rocm3.3-9ca344d` docker to run on a gfx906. It does not abort but got frozen for hours. I managed to have a VLOG=3 dump. Last three lines of log indicates it got frozen in ThenMemcpy.

`2020-07-08 13:58:51.529967: I tensorflow/core/common_runtime/gpu/gpu_util.cc:303] CopyCPUTensorToGPU
2020-07-08 13:58:51.529979: I tensorflow/stream_executor/stream.cc:2205] [stream=0x7e44c50,impl=0x7e441f0] Called Stream::ThenWaitFor(other=0x7e38e30)
2020-07-08 13:58:51.544605: I tensorflow/stream_executor/stream.cc:5155] [stream=0x7e44c50,impl=0x7e441f0] Called Stream::ThenMemcpy(gpu_dst=0x7f1df6600600, host_src=0x8069b40, size=4)`

**Comment by MJDSys** [comment]
I'm trying to compile my own TF from source (using the develop-upstream branch) to see if that fixes anything, but I ran into some build issues.

In the meantime, I did play around with a debugger.  I managed to check what error code is being passed to callbackQueue (the function that is actually aborting), and it appears to be an HSA error ILLEGAL_INSTRUCTION.  Let me know if you need any more information.

I'm also looking to see if I can move my current in production docker image back to ROCm 3.3 while this bug is being fixed.  If that does fix it, is there any easy way to compile the entire ROCm stack?  I could try git bisect'ing this issue if so.

**Comment by MJDSys** [comment]
I got my build of this repository working against 3.5.1 using the develop-upstream branch, and I can confirm that stopped the test script (and my actual production code) from crashing.

If that's enough to consider this issue fixed, I'm good for now as I have two potential workarounds (master branch/ROCm 3.3).  Please manage it as needed, and let me know if any more information would be helpful.

**Comment by jerryyin** [comment]
@MJDSys Thanks for the information. Just to confirm I understand it right:

- The script crashes on ROCm 3.5.1; Using the latest `develop-upstream`
- The script runs to get to correct result on ROCm 3.3; Using the latest `develop-upstream`?

**Comment by MJDSys** [comment]
@jerryyin Not quite.  What I've tested:
 * Using ROCm 3.5.0/3.5.1 and the released tensorflow-rocm PyPi package, the script crashes.  This happens with both my custom docker image and the official docker image released by AMD.
 * Using ROCm 3.5.1 using the latest develop-upstream, the script runs to completion.

This affects both the script posted here (a minimal reproducer) along with my actual models.

**Comment by jerryyin** [comment]
Thanks. We are still triaging this issue, but it will be fixed in the coming `tensorflow-rocm-2.3-enhanced` release according to @sunway513. I'm keeping the issue open when a docker image become publicly available.

**Comment by jerryyin** [comment]
Could you give `docker pull rocm/tensorflow:rocm3.8-tf2.3-dev` a shot? Thanks

**Comment by jerryyin** [comment]
Closing as fixed
